### PR TITLE
Allow explicitly plumbing through nics

### PR DESCRIPTION
**Pull request by tgaddair**
Allows us to circumvent the `detect_nics` step, which is not necessary in all environments.

### CODE REVIEW

**Comment by tgaddair** [comment]
Going to add some tests before we land this.

**Comment by arnavgarg1** [comment]
> Going to add some tests before we land this.

PR looks good to me, will wait for tests to be added!

File: ludwig/backend/ray.py [hunk 1]
```diff
@@ -331,6 +330,31 @@ def process_results(self, results: List[Dict], **info) -> None:
...
+            # When doing CPU-only training, default to a SPREAD policy to avoid
+            # packing too many workers on a single machine
+            os.environ[TRAIN_ENABLE_WORKER_SPREAD_ENV] = "1"
+
+        trainer = Trainer(**trainer_kwargs)
+    finally:
+        if prev_env is not None:
+            os.environ[TRAIN_ENABLE_WORKER_SPREAD_ENV] = prev_env
+        elif TRAIN_ENABLE_WORKER_SPREAD_ENV in os.environ:
+            del os.environ[TRAIN_ENABLE_WORKER_SPREAD_ENV]
```
[the remark this answers is not in the document]
Comment (tgaddair) [reply ludwig/backend/ray.py]: For the delete? Should be safe based on the above check.

File: tests/conftest.py [hunk 2]
```diff
@@ -24,6 +25,19 @@
...
+def pytest_sessionstart(session):
+    session.start_time = time.time()
+
+
+@pytest.fixture(autouse=True)
+def check_session_time(request):
+    elapsed = time.time() - request.session.start_time
+    if elapsed > TEST_SUITE_TIMEOUT_S:
+        request.session.shouldstop = "time limit reached: %0.2f seconds" % elapsed
+
```
Comment (arnavgarg1) [inline tests/conftest.py:40]: Nice!

**Comment by arnavgarg1** [comment]
Also noticed that the tests failed with error code 2 (different than normal). Is that because of the explicit timer that's added to conftest?

**Comment by tgaddair** [comment]
@arnavgarg1 the NIC issue comes up because sometimes the network interface picked by MPI or Gloo might be the fastest or might be invalid. For example, we typically need to ignore `lo` and `docker` as these are not routable. In some cases, there may be other interfaces that are listed but unroutable. Ideally you want to find a common set among all the workers.

This came up recently because of https://github.com/horovod/horovod/issues/1855, where Horovod's compatible network interface detection logic was failing because of the presence of the Istio NAT. It looks like in cases where there is a pretty standard interface setup among the workers, just setting the NIC to the empty string allows Gloo to pick what it thinks is the best NIC and this works well enough more often than not.

Regarding the error code, yes, that's the error code returned from pytest when it times out with the new logic.

**Pull request merged**